### DISCUSSION TITLE
Mutating variables breaks type safety

### DISCUSSION THREAD

**Discussion by kishaningithub** [cue-lang/cue#1870] [Q&A] [answered]
Consider this cue file

```cue
join: [...int]
join: [1, 2, ...]
join: join + [3, 4, "some string"]
```

Exported json

```json
{
    "join": [
        1,
        2,
        3,
        4,
        "some string"
    ]
}
```

Playground link: https://cuelang.org/play/?id=cgbhUBpEUwA#cue@export@json

If you see above, in first line i am defining join as an open array of integers but in line 3 i am adding a string, this does not throw an error.

Is this a feature or a bug?

**Comment by verdverm** [comment, marked as answer]
I don't know the answer to your question, but generally, mutation is not permitted in CUE.


when experimenting, the following (from tip) seems like a bug to me

```sh
cue eval in.cue
-- in.cue --
a: a

-- out.txt --
a: _
```

I would have expected an error, first that the RHS `a` is a bad reference, second a cycle. I get the same output on 0.4.3, noting that I have the latest cycle detection reimplementation on tip in the original.

Another example where I would expect a cycle error rather than conflicting values

```cue
import "strings"
a: "A"
a: strings.ToLower(a)
```

**Reply by kishaningithub** [reply]
Have created a bug to track this https://github.com/cue-lang/cue/issues/1874

**Reply by myitcv** [reply]
I'm not clear that:

```
a: a
```

resulting in:

```
a: _
``` 

is a bug. Because it's simply constraining `a` to be the value of `a`. Yes this is a tautology, but not wrong in the result or semantics. The only thing that declaration does is say that `a` is defined. But we know nothing more about `a`, so the result of `_` makes sense to me. That result is by definition incomplete, which is also correct.

Therefore, I'll close #1874 because there is no mutation being allowed.

@verdverm's example is more interesting. If nothing else, I think we to ensure the spec is clear on such an example. I'll raise an issue for that.

**Reply by myitcv** [reply]
I should have been more precise when referring to @verdverm's example. I was referring to this one as being interesting:

```cue
import "strings"
a: "A"
a: strings.ToLower(a)
```

**Reply by myitcv** [reply]
https://github.com/cue-lang/cue/issues/1880

**Reply by mpvl** [reply]
Indeed:
```
a: a
```
is valid, logically, as it just means that `a` must be equal to `a`. and is equivalent to `a: _`.

This can be used to ensure that to values are mutually equal:

```
a: b
b: a
```
or even 
```
a: b + 100
b: a  - 100
```
This is all valid and documented in the spec.

Something like `a: a + 1`, however, is not valid, as a value cannot be itself plus 1. The OP's example is of the latter kind.

**Comment by antong** [comment]
It looks like a bug to me. It is not mutation, and the self reference is not disallowed (re the previous answer.) But, I don't see how the exported data could have a string in a list of ints. Even shorter example:
```
x: [...int]
x: x + ["foo"]
// -> x: ["foo"]
```
The exported output doesn't satisfy the input definition (the exported output doesn't unify with the definition.)

The following small amendment triggers the (expected) error:
```
x: [...int]
x: x + ["foo"]
_a: x
// -> conflicting values "foo" and int (mismatched types string and int)
```

**Reply by kishaningithub** [reply]
@myitcv Does your PR https://github.com/cue-lang/cue/issues/1880 fix this as well?

**Reply by antong** [reply]
#1898

**Comment by mpvl** [comment]
It seems to me that this should fail. I'm working on some fixes on comprehensions and late evaluation in self references (like here), where this would fail.
